x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanStreamInput.java

```diff
@@ -206,4 +210,17 @@ static void throwOnNullOptionalRead(Class<?> type) throws IOException {
     public NameId mapNameId(long l) {
         return nameIdFunction.apply(l);
     }
+
+    public FieldAttribute attributeFromCache(int id) {
+        assert fieldAttributes.length > id && fieldAttributes[id] != null;
+        return fieldAttributes[id];
+    }
+
+    public void toCache(FieldAttribute attr, int id) {
```

Comment:
bikeshed: wouldn't it be more natural to have the cache generate the id for you and return it? Then we'd also guarantee that the ids are actually contiguous.

Reply:
These IDs are generated from the coordinator node. 
I could just increment a counter here, but it doesn't make me comfortable, I really prefer to stick to the IDs I get from the network and that I'll have to use later to retrieve the Attributes

Reply:
When I first read it I had a similar thought. And then I realized why it's the way it is. Maybe we're best off putting a comment saying that the cache comes from reading.